dbms/src/Storages/KVStore/MultiRaft/Disagg/FastAddPeer.cpp

```diff
@@ -246,136 +172,292 @@ bool tryResetPeerIdInRegion(RegionPtr region, const RegionLocalState & region_st
+FastAddPeerRes FastAddPeerImpl(
+    FastAddPeerContextPtr fap_ctx,
+    TMTContext & tmt,
+    TiFlashRaftProxyHelper * proxy_helper,
+    UInt64 region_id,
+    UInt64 new_peer_id,
+    UInt64 start_time)
+{
+    try
+    {
+        auto elapsed = fap_ctx->tasks_trace->queryElapsed(region_id);
+        GET_METRIC(tiflash_fap_task_duration_seconds, type_queue_stage).Observe(elapsed / 1000.0);
+        GET_METRIC(tiflash_fap_task_state, type_queueing_stage).Decrement();
+        auto res = FastAddPeerImplSelect(tmt, proxy_helper, region_id, new_peer_id);
+        if (std::holds_alternative<CheckpointRegionInfoAndData>(res))
+        {
+            auto final_res = FastAddPeerImplWrite(
+                tmt,
+                region_id,
+                new_peer_id,
+                std::move(std::get<CheckpointRegionInfoAndData>(res)),
+                start_time);
+            GET_METRIC(tiflash_fap_task_result, type_success_transform).Increment();
+            return final_res;
+        }
+        return std::get<FastAddPeerRes>(res);
+    }
+    catch (const Exception & e)
+    {
+        // Could be like:
+        // - can't put remote page with empty data_location
+        DB::tryLogCurrentException(
+            "FastAddPeerImpl",
+            fmt::format(
+                "Failed when try to restore from checkpoint region_id={} new_peer_id={} {}",
+                region_id,
+                new_peer_id,
+                e.message()));
+        GET_METRIC(tiflash_fap_task_result, type_failed_baddata).Increment();
+        return genFastAddPeerRes(FastAddPeerStatus::BadData, "", "");
     }
     catch (...)
     {
-        DB::tryLogCurrentException("FastAddPeer", "Failed when try to restore from checkpoint");
+        DB::tryLogCurrentException(
+            "FastAddPeerImpl",
+            fmt::format(
+                "Failed when try to restore from checkpoint region_id={} new_peer_id={}",
+                region_id,
+                new_peer_id));
+        GET_METRIC(tiflash_fap_task_result, type_failed_baddata).Increment();
         return genFastAddPeerRes(FastAddPeerStatus::BadData, "", "");
     }
 }
 
+void ApplyFapSnapshotImpl(TMTContext & tmt, TiFlashRaftProxyHelper * proxy_helper, UInt64 region_id, UInt64 peer_id)
+{
+    auto * log = &Poco::Logger::get("FastAddPeer");
+    LOG_INFO(log, "Begin apply fap snapshot, region_id={}, peer_id={}", region_id, peer_id);
+    GET_METRIC(tiflash_fap_task_state, type_ingesting_stage).Increment();
+    SCOPE_EXIT({ GET_METRIC(tiflash_fap_task_state, type_ingesting_stage).Decrement(); });
+    Stopwatch watch_ingest;
+    auto kvstore = tmt.getKVStore();
+    auto fap_ctx = tmt.getContext().getSharedContextDisagg()->fap_context;
+    auto checkpoint_ingest_info = fap_ctx->getOrRestoreCheckpointIngestInfo(tmt, proxy_helper, region_id, peer_id);
+    kvstore->handleIngestCheckpoint(checkpoint_ingest_info->getRegion(), checkpoint_ingest_info, tmt);
+    checkpoint_ingest_info->markDelete();
```

Comment:
I think calling `checkpoint_ingest_info->removeFromLocal` explicitly here is more reasonable, why do you only mark it as delete here?

Reply:
I think the deleting phase could be done asynchronously outside the `ApplyFapSnapshotImpl` call to prevent blocking the region-worker on proxy's side for too long.
However, this optimization is not implemented in this pr.